File: core/src/main/java/com/lomekwi/cave/util/Duplicatable.java
package com.lomekwi.cave.util;

import java.io.*;

public interface Duplicatable<T extends Duplicatable<T>> extends Serializable {

    @SuppressWarnings("unchecked")
    /*
     * 创建一个深拷贝
     */
    default T duplicate() {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
             ObjectOutputStream oos = new ObjectOutputStream(baos)) {
            oos.writeObject(this);
            try (ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(baos.toByteArray()))) {
                return (T) ois.readObject();
            }
        } catch (IOException | ClassNotFoundException e) {
            throw new RuntimeException("Duplicate failed", e);
        }
    }
}

File: core/src/main/scala/com/lomekwi/cave/util/Duplicatable.scala
package com.lomekwi.cave.util

import java.io.*
import scala.util.Using

trait Duplicatable[T <: Duplicatable[T]] extends Serializable {

  /*
   * 创建一个深拷贝
   */
  def duplicate(): T = {
    try {
      Using.resource(new ByteArrayOutputStream()) { baos =>
        Using.resource(new ObjectOutputStream(baos)) { oos =>
          oos.writeObject(this)
          Using.resource(new ObjectInputStream(new ByteArrayInputStream(baos.toByteArray))) { ois =>
            ois.readObject().asInstanceOf[T]
          }
        }
      }
    } catch {
      case e: IOException => throw new RuntimeException("Duplicate failed", e)
      case e: ClassNotFoundException => throw new RuntimeException("Duplicate failed", e)
    }
  }
}
